test/e2e/framework/service_util.go

```diff
@@ -1450,3 +1461,75 @@ func GetServiceLoadBalancerCreationTimeout(cs clientset.Interface) time.Duration
 	}
 	return LoadBalancerCreateTimeoutDefault
 }
+
+func TestAffinityFromPod(execPod *v1.Pod, targetIpPort string, shouldHold bool) bool {
+	var hostname string
+	cmd := fmt.Sprintf(`wget -qO- http://%s/ -T 2`, targetIpPort)
+	affinityCount := 0
+	if pollErr := wait.PollImmediate(Poll, ServiceTestTimeout, func() (bool, error) {
+		stdout, err := RunHostCmd(execPod.Namespace, execPod.Name, cmd)
+		if err != nil || len(stdout) == 0 {
+			Logf("Failed to get response from %s. Retry until timeout", targetIpPort)
+			hostname = ""
+			affinityCount = 0
+			return false, nil
+		}
+		Logf("Got pod response: %s", stdout)
+		if shouldHold {
+			if hostname == stdout {
+				affinityCount++
+			} else {
+				affinityCount = 0
+				hostname = stdout
+			}
+			return affinityCount >= AffinityConfirmCount, nil
+		} else {
+			if hostname == "" {
+				hostname = stdout
+				return false, nil
+			} else {
+				return hostname != stdout, nil
+			}
+		}
+	}); pollErr != nil {
+		// Affinity check timed out.
+		return false
+	}
+	return true
+}
+
+func TestAffinityFromExternal(jig *ServiceTestJig, targetIp string, targetPort int, shouldHold bool) bool {
```

Comment:
Combine this with `TestAffinityFromPod` with one more knob? Looks pretty similar.